src/solaris.rs

```diff
@@ -0,0 +1,28 @@
+//! Solaris implementation using getrandom(2).
+//!
+//! While getrandom(2) has been available since Solaris 11.3, it has a few
+//! quirks not present on other OSes. First, on Solaris 11.3, calls will always
+//! fail if bufsz > 1024. Second, it will always either fail or completely fill
+//! the buffer (returning bufsz). Finally, "if GRND_RANDOM is not specified
+//! then getrandom(2) is always a non blocking call". This _might_ imply that
+//! in early-boot scenarios with low entropy, getrandom(2) will not properly
+//! block. To be safe, we set GRND_RANDOM, mirroring the man page examples.
+//!
+//! For more information, see the man page linked in lib.rs and this blog post:
+//! https://blogs.oracle.com/solaris/post/solaris-new-system-calls-getentropy2-and-getrandom2
+//! which also explains why this crate should not use getentropy(2).
+use crate::{util_libc::last_os_error, Error};
+use core::mem::MaybeUninit;
+
+const MAX_BYTES: usize = 1024;
+
+pub fn getrandom_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
+    for chunk in dest.chunks_mut(MAX_BYTES) {
+        let ptr = chunk.as_mut_ptr() as *mut libc::c_void;
+        let ret = unsafe { libc::getrandom(ptr, chunk.len(), libc::GRND_RANDOM) };
+        if ret != (chunk.len() as isize) {
+            return Err(last_os_error());
+        }
```

Comment:
It also may be worth to add a note that `getrandom_inner` never receives an empty buffer.

Reply:
Fixed, but I made the error check `ret < 0`, as it's negative values which indicates errors for `getrandom`.

Reply:
Oh weird it seems like the docs contradict themselves:

"Upon successful completion, the getrandom() function returns the number of bytes written to buf. Otherwise, it returns 0 and sets errno to indicate the error."

"the function returns -1 and errno is set to EAGAIN."

Reply:
No, the check should be `ret == 0` as per the cited man quote (yeah, I was surprised as well...).

Reply:
I think checking for both negative and zero return values is probably the safe thing to do here. I'm shocked how weird the solaris implementation of getrandom is, considering how normal the Illumos one is.

Reply:
>the function returns -1 and errno is set to EAGAIN.

IIUC it only applies to `GRND_NONBLOCK`.

Reply:
Updated to check for negative and zero values.

Reply:
It could be worth to slightly modify the check to have only one branch in the happy path:
```rust
if ret as usize != chunk.len() {
    return if ret == 0 || ret == -1 {
        last_os_error()
    } else {
        Err(Error::UNEXPECTED)
    };
}
```

Reply:
I think that is harder to read (and will probably get compiled to the same thing regardless).

Reply:
You can tweak formatting to your liking.

>will probably get compiled to the same thing regardless

It does not: https://rust.godbolt.org/z/G6d1zq4hs But it's a small enough difference, so I do not insist on this change.

Reply:
Oh interesting. I'll leave it as-is, but that's good to know.